src/routes/_surfaces/workspace/explore/time-controls/time-range-utils.ts

```diff
@@ -0,0 +1,548 @@
+// For custom time ranges, we'll need another function with "breakpoint" logic that analyzes the user-determined start/end dates.
+export const getSelectableTimeGrains = (
+  timeRangeName: TimeRangeName,
+  allTimeRange: TimeSeriesTimeRange
+): TimeGrain[] => {
+  if (!allTimeRange) return [];
+  const timeRangeDuration = getTimeRangeDuration(timeRangeName, allTimeRange);
+
+  const timeGrains: TimeGrain[] = [];
+  for (const timeGrain in TimeGrain) {
+    // only show a time grain if it results in a reasonable number of points on the line chart
+    const MINIMUM_POINTS_ON_LINE_CHART = 2;
+    const MAXIMUM_POINTS_ON_LINE_CHART = 2500;
+    const timeGrainDuration = getTimeGrainDuration(TimeGrain[timeGrain]);
+    const pointsOnLineChart = timeRangeDuration / timeGrainDuration;
+    const showTimeGrain =
+      pointsOnLineChart >= MINIMUM_POINTS_ON_LINE_CHART &&
+      pointsOnLineChart <= MAXIMUM_POINTS_ON_LINE_CHART;
+    if (showTimeGrain) {
+      timeGrains.push(TimeGrain[timeGrain]);
+    }
+  }
+  if (timeGrains.length === 0) {
+    throw new Error(`No time grains generated for time range ${timeRangeName}`);
+  }
+  return timeGrains;
+};
+
+export const getDefaultTimeGrain = (
+  timeRangeName: TimeRangeName
+): TimeGrain => {
+  switch (timeRangeName) {
+    // case TimeRangeName.LastHour:
+    //   return TimeGrain.FifteenMinutes;
+    case TimeRangeName.Last6Hours:
+      return TimeGrain.OneHour;
+    case TimeRangeName.LastDay:
+      return TimeGrain.OneHour;
+    case TimeRangeName.Last2Days:
+      return TimeGrain.OneHour;
+    case TimeRangeName.Last5Days:
+      return TimeGrain.OneHour;
+    case TimeRangeName.LastWeek:
+      return TimeGrain.OneHour;
+    case TimeRangeName.Last2Weeks:
+      return TimeGrain.OneDay;
+    case TimeRangeName.Last30Days:
+      return TimeGrain.OneDay;
+    case TimeRangeName.Last60Days:
+      return TimeGrain.OneDay;
+    case TimeRangeName.AllTime:
+      // TODO: this needs breakpoint logic using start/end time.
+      return TimeGrain.OneDay;
+    default:
+      throw new Error(`No default time grain for time range ${timeRangeName}`);
+  }
+};
+
+export const makeTimeRange = (
+  timeRangeName: TimeRangeName,
+  timeGrain: TimeGrain,
+  allTimeRange: TimeSeriesTimeRange
+): TimeSeriesTimeRange => {
+  switch (timeRangeName) {
+    // case TimeRangeName.LastHour: {
+    //   const anchorDate = roundDateUp(new Date(allTimeRange?.end), timeGrain);
+    //   const startDate = new Date(anchorDate.getTime() - 60 * 60 * 1000);
+    //   const endDate = subtractOneUnit(anchorDate, TimeUnit.Second);
+    //   return {
+    //     name: timeRangeName,
+    //     start: startDate.toISOString(),
+    //     end: endDate.toISOString(),
+    //     interval: timeGrain.toString(),
+    //   };
+    // }
+    case TimeRangeName.Last6Hours: {
+      const anchorDate = roundDateUp(new Date(allTimeRange?.end), timeGrain);
```

Comment:
Can we get rid of all the common code here? Create a function that takes the variables.

Reply:
Yep, just refactored it. A nice little step towards supporting more kinds of time ranges (like Fiscal periods, or X-to-Date)